go/vt/vtgate/engine/distinct.go

```diff
@@ -105,32 +113,42 @@ func (pt *probeTable) hashCodeForRow(inputRow row) (evalengine.HashCode, error)
 	// > So you want an odd MULT that has plenty of one-bits.
 
 	code := evalengine.HashCode(17)
-	for idx, value := range inputRow {
-		// We use unknown collations when we do not have collation information
-		// This is safe for types which do not require collation information like
-		// numeric types. It will fail at runtime for text types.
-		collation := collations.Unknown
-		if len(pt.colCollations) > idx {
-			collation = pt.colCollations[idx]
+	for i, checkCol := range pt.checkCols {
+		if i >= len(inputRow) {
+			return 0, vterrors.New(vtrpcpb.Code_INTERNAL, "distinct check colls is larger than its input row")
 		}
-		hashcode, err := evalengine.NullsafeHashcode(value, collation, value.Type())
+		col := inputRow[checkCol.Col]
+		hashcode, err := evalengine.NullsafeHashcode(col, checkCol.Collation, col.Type())
 		if err != nil {
-			return 0, err
+			if err != evalengine.UnsupportedCollationHashError || checkCol.WsCol == nil {
+				return 0, err
+			}
+			checkCol = checkCol.SwitchToWeightString()
+			pt.checkCols[i] = checkCol
+			hashcode, err = evalengine.NullsafeHashcode(inputRow[checkCol.Col], checkCol.Collation, col.Type())
+			if err != nil {
+				return 0, err
+			}
```

Comment:
Should we not do this the other way around? Like check that `checkCol.WsCol` is not nil and use it. If it is nil, then we do the normal hash. In this case we won't have to check for the Unsupported collation error

Reply:
Nah, if the column is numerical, we can't use the `weight_string` value and hashing works just fine. We could have checked the type here, but that is kind of what the evalengine is already doing, so I thought this would consolidate the type checking to fewer places